httpclient5/src/main/java/org/apache/hc/client5/http/impl/nio/HappyEyeballsV2AsyncClientConnectionOperator.java

```diff
@@ -0,0 +1,698 @@
+     * @throws IllegalArgumentException if {@code tlsStrategyLookup} is {@code null}.
+     */
+    public HappyEyeballsV2AsyncClientConnectionOperator(
+            final Lookup<TlsStrategy> tlsStrategyLookup) {
+        this(tlsStrategyLookup, DefaultSchemePortResolver.INSTANCE, null);
+    }
+
+
+    /**
+     * Attempts to connect to the given host and returns a Future that will be completed when the connection is established
+     * or when an error occurs. This method may attempt to connect to multiple IP addresses associated with the host,
+     * depending on the address family and the number of connection attempts to execute. The address family and number of
+     * connection attempts can be configured by calling the corresponding setters on this class.
+     *
+     * @param connectionInitiator the connection initiator to use when creating the connection
+     * @param host                the host to connect to
+     * @param localAddress        the local address to bind to when connecting, or null to use any available local address
+     * @param connectTimeout      the timeout to use when connecting, or null to use the default timeout
+     * @param attachment          the attachment to associate with the connection, or null if no attachment is needed
+     * @param callback            the callback to invoke when the connection is established or an error occurs, or null if no callback is needed
+     * @return a Future that will be completed when the connection is established or when an error occurs
+     */
+    @Override
+    public Future<ManagedAsyncClientConnection> connect(
+            final ConnectionInitiator connectionInitiator,
+            final HttpHost host,
+            final SocketAddress localAddress,
+            final Timeout connectTimeout,
+            final Object attachment,
+            final FutureCallback<ManagedAsyncClientConnection> callback) {
+
+        final CompletableFuture<ManagedAsyncClientConnection> connectionFuture = new CompletableFuture<>();
+
+        final Timeout conTimeout = connectTimeout != null ? connectTimeout : timeout;
+
+        resolveDnsAsync(host.getHostName())
+                .thenCompose(inetAddresses -> {
+                    final List<InetAddress> ipv4Addresses = new ArrayList<>();
+                    final List<InetAddress> ipv6Addresses = new ArrayList<>();
+
+                    for (final InetAddress inetAddress : inetAddresses) {
+                        if (inetAddress instanceof Inet4Address) {
+                            ipv4Addresses.add(inetAddress);
+                        } else if (inetAddress instanceof Inet6Address) {
+                            ipv6Addresses.add(inetAddress);
+                        }
+                    }
+
+                    // Sort the array of addresses using the custom Comparator
+                    Arrays.sort(inetAddresses, InetAddressComparator.INSTANCE);
```

Comment:
Is this a [stable sort](https://www.geeksforgeeks.org/stable-and-unstable-sorting-algorithms/)? It's important to not change the order of IP addresses that are equal under the comparison rules.